INT-4243: Upgrade to sshd 1.4

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4243

Remove newline char from key exchange.

__we could consider just backporting the newline stripping if updating the test server version will cause issues with IO__

garyrussell committed:
INT-4243: Upgrade to sshd 1.4

JIRA: https://jira.spring.io/browse/INT-4243

Remove newline char from key exchange.

artembilan commented on spring-integration-sftp/src/test/java/org/springframework/integration/sftp/session/SftpServerTests.java:
```diff
@@ -125,7 +127,12 @@ private void testKeyExchange(String pubKey, String privKey, String passphrase)
-		byte[] decodeBuffer = Base64.decodeBase64(StreamUtils.copyToByteArray(stream));
+		byte[] keyBytes = StreamUtils.copyToByteArray(stream);
+		// strip any newline chars
+		while (keyBytes[keyBytes.length - 1] == 0x0a || keyBytes[keyBytes.length - 1] == 0x0c) {
```
Any ideas why that failed sporadically?
Thanks

Merging BTW

garyrussell replied:
See my comment on the JIRA - problem is unrelated but we should do the upgrade anyway.

artembilan commented on spring-integration-sftp/src/test/java/org/springframework/integration/sftp/SftpTestSupport.java:
```diff
@@ -60,9 +62,10 @@ public static void createServer() throws Exception {
+		server.setKeyPairProvider(new SimpleGeneratorHostKeyProvider(new File("hostkey.ser")));
+		server.setSubsystemFactories(Collections.<NamedFactory<Command>>singletonList(new SftpSubsystemFactory()));
+		server.setFileSystemFactory(
+				new VirtualFileSystemFactory(Paths.get(remoteTemporaryFolder.getRoot().getAbsolutePath())));
```
Why `remoteTemporaryFolder.getRoot().toPath()` doesn't work for you?

garyrussell replied:
ok; fine

artembilan commented on spring-integration-sftp/src/test/java/org/springframework/integration/sftp/session/SftpServerTests.java:
```diff
@@ -125,7 +127,12 @@ private void testKeyExchange(String pubKey, String privKey, String passphrase)
+		while (keyBytes[keyBytes.length - 1] == 0x0a || keyBytes[keyBytes.length - 1] == 0x0c) {
+			keyBytes = Arrays.copyOf(keyBytes, keyBytes.length - 1);
+		}
+		byte[] decodeBuffer = Base64Utils.decode(keyBytes);
```
Hm... Fails for me on Windows like this:
```
java.lang.IllegalArgumentException: Illegal base64 character d
	at java.util.Base64$Decoder.decode0(Base64.java:714)
	at java.util.Base64$Decoder.decode(Base64.java:526)
	at org.springframework.util.Base64Utils.decode(Base64Utils.java:63)
	at org.springframework.integration.sftp.session.SftpServerTests.decodePublicKey(SftpServerTests.java:135)
```

garyrussell replied:
Oops, my mistake should be 0x0d not ox0c

garyrussell committed:
Fix CR and Add Windows Delete Diagnostics

garyrussell committed:
Fix For Windows

More diagnostics.

It appears the new server doesn't close the file when the session is closed.

In the streaming test, consume the stream before closing the session.

garyrussell committed:
Polishing - PR Comments

artembilan commented on spring-integration-sftp/src/test/java/org/springframework/integration/sftp/dsl/SftpTests.java:
```diff
@@ -24,11 +24,13 @@
 import java.util.List;
 import java.util.regex.Matcher;
 
+import org.apache.log4j.lf5.util.StreamUtils;
```
??? Let me guess you were going to use `org.springframework.util.StreamUtils`

garyrussell replied:
hehe

artembilan commented on spring-integration-sftp/src/test/java/org/springframework/integration/sftp/dsl/SftpTests.java:
```diff
@@ -117,12 +119,14 @@ public void testSftpInboundStreamFlow() throws Exception {
 		assertNotNull(message);
 		assertThat(message.getPayload(), instanceOf(InputStream.class));
 		assertThat(message.getHeaders().get(FileHeaders.REMOTE_FILE), isOneOf(" sftpSource1.txt", "sftpSource2.txt"));
+		StreamUtils.copy((InputStream) message.getPayload(), new ByteArrayOutputStream());
```
Well, my solution is actually like this:
```
((InputStream) message.getPayload()).close();
```
Because the problem with the file on Windows that we have an open stream to it. And, yeah..., exactly payload is `InputStream` in case of streaming inbound channel adapter.

garyrussell replied:
ok; feel free to fix that way.

garyrussell replied:
Interestingly, windows did not say that the file deletion failed, just silently left the file in place, preventing the deletion of the directory.

artembilan replied:
Well, actually it does, if we check the result of the `File.delete()`.
And that is exactly what we are doing now with all those your new log diagnostics.

OK, merging having your agreement...
I also will increase the timeout in that Redis test

garyrussell replied:
Right that's what I am saying; on my Windows VM, the file.delete() was successful (I didn't get the "couldn't delete" message - that's why I added the additional "Contents" log when the directory delete fails.

artembilan commented on spring-integration-file/src/test/java/org/springframework/integration/file/remote/RemoteFileTestSupport.java:
```diff
@@ -162,15 +172,24 @@ public void recursiveDelete(File file) {
 				}
 			}
-			file.delete();
+			logger.info("Deleting: " + file + " in " + testName.getMethodName());
```
I think we have figure out the problem with files and we don't need these diagnostics any more.
Let me know and I'll polish on merge respectively.

Thank you for the fix any way!

garyrussell replied:
I figured it won't hurt leaving these at INFO level - in case we have problems in future.

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/4bb3f5041f0a84cd736399fc1a0da96b7aeab742 and cherry-picked to `4.3.x` as https://github.com/spring-projects/spring-integration/commit/96331d5415dd4b6b04bd5bd683884d57aa5848fe

artembilan commented:
Back-ported to `4.2.x` as https://github.com/spring-projects/spring-integration/commit/8036a4bab14082a56b6e0b24a195165a15659140
